very-heavy-job/src/main/java/org/example/StreamingJob.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.example;
+
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.streaming.api.functions.source.FromSplittableIteratorFunction;
+import org.apache.flink.util.Collector;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.util.Arrays;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Skeleton for a Flink Streaming Job.
+ *
+ * <p>For a tutorial how to write a Flink streaming application, check the
+ * tutorials and examples on the <a href="https://flink.apache.org/docs/stable/">Flink Website</a>.
+ *
+ * <p>To package your application into a JAR file for execution, run
+ * 'mvn clean package' on the command line.
+ *
+ * <p>If you change the name of the main class (with the public static void main(String[] args))
+ * method, change the respective entry in the POM.xml file (simply search for 'mainClass').
+ */
+public class StreamingJob {
+
+	public static void main(String[] args) throws Exception {
+		// set up the streaming execution environment
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		env.enableCheckpointing(TimeUnit.SECONDS.toMillis(10));
+		env.getCheckpointConfig().setCheckpointTimeout(TimeUnit.MINUTES.toMillis(1));
+
+		final int initialBytesSize = 1024 * 1024 ; // 1Mb
+		final int stateSize = (int) (1024L * 1024L * 1024L * 12L / (256L * 256L)); // 12Gb in total across all keys
+		final int outputSize = 1024 * 1024 * 10 ; // 10Mb
+
+		// execute program
+		env
+				.addSource(new FromSplittableIteratorFunction<>(new RandomSplittableIterator(initialBytesSize)), TypeInformation.of(byte[].class))
+				.keyBy(record -> (((int)record[0]) * 256) + (int)record[1]) // Use the first byte as key
+				.process(new StateAllocatorFunction(stateSize, outputSize), TypeInformation.of(byte[].class))
+				.addSink(new SinkFunction<byte[]>() {
+					@Override
+					public void invoke(byte[] value, Context context) throws Exception {
+						SinkFunction.super.invoke(value, context);
+					}
+				});
+
+		env.execute("Very heavy job");
+	}
+
+
+	public static class StateAllocatorFunction
+			extends KeyedProcessFunction<Integer, byte[], byte[]> {
```

Comment:
nit: Instead of sending byte[] we could also send random integers or so that we restrict in its range. This might be a bit simpler.